python/ray/experimental/channel/shared_memory_channel.py

```diff
@@ -480,6 +484,71 @@ def close(self) -> None:
         self._worker.core_worker.experimental_channel_set_error(self._reader_ref)
 
 
+@DeveloperAPI
+class BufferedSharedMemoryChannel(ChannelInterface):
+    """A Channel that allow to buffer inputs.
+
+    The write/read APIs are non-blocking as long as the buffers are not
+    exhausted. The caller of the API should guarantee to consume buffers
+    before they are exhausted, otherwise write/read raises
+    `RayChannelTimeoutError`.
+    """
+
+    def __init__(
+        self,
+        writer: Optional[ray.actor.ActorHandle],
+        reader_and_node_list: List[Tuple["ray.actor.ActorHandle", str]],
+        num_shm_buffers: int,
+        typ: Optional[Union[int, SharedMemoryType]] = None,
+    ):
+        self._num_shm_buffers = num_shm_buffers
+        self._channels = [
+            Channel(writer, reader_and_node_list, typ) for _ in range(num_shm_buffers)
+        ]
+        self._next_write_index = 0
+        self._next_read_index = 0
+
+    def ensure_registered_as_writer(self):
+        """
+        Check whether the process is a valid writer. This method must be idempotent.
+        """
+        for chan in self._channels:
+            chan.ensure_registered_as_writer()
+
+    def ensure_registered_as_reader(self):
+        """
+        Check whether the process is a valid reader. This method must be idempotent.
+        """
+        for chan in self._channels:
+            chan.ensure_registered_as_reader()
+
+    def write(self, value: Any, timeout: Optional[float] = None) -> None:
+        self._channels[self._next_write_index].write(value, timeout)
+        self._next_write_index += 1
+        self._next_write_index %= self._num_shm_buffers
+
+    def read(self, timeout: Optional[float] = None) -> Any:
+        output = self._channels[self._next_read_index].read(timeout)
+        self._next_read_index += 1
```

Comment:
ditto: what happens when the index wraps around? Could wraparound happen? Add docstring